### PR TITLE
R2l

### PR DESCRIPTION
**Pull request by rodsol**
Adverbial preposition
Example:  "we shall meet at the library."
   _advmod(meet, at)
   _pobj(at, library)

==>

(EvaluationLink 
  (PredicateNode "at@123")
                (ListLink 
                          (PredicateNode "meet@456") 
                          (ConceptNode "library@789")))

### CODE REVIEW

**Comment by williampma** [comment]
I don't think that's the correct use of SatisfyingSetLink.  Not according to http://wiki.opencog.org/w/SatisfyingSetLink anyway.  What is a satisfying set with no variable?

Though I seem to recall there was some unconcluded discussion about it.

File: opencog/nlp/relex2logic/rule-helpers.scm [hunk 1]
```diff
@@ -419,6 +419,19 @@
 	(InheritanceLink  (SatisfyingSetLink (PredicateNode instance)) (ConceptNode adv_instance))
 	)
 )
+
+(define (adverbialpp-rule verb instance prep  prep_instance noun noun_instance)
+    (list (InheritanceLink  (ConceptNode noun_instance) (ConceptNode noun))
```
Comment (AmeBel) [inline opencog/nlp/relex2logic/rule-helpers.scm]: There is no rule on relex that invokes this funtion

**Comment by AmeBel** [comment]
@anitzkin greetings. Any comment on this? Rodas is reviewing your changes, and trying to understand what is going on. Your assistance is greatly appreciated.

**Comment by anitzkin** [comment]
Yeah, I took a glance at this stuff earlier.  I will have some comments,
for sure; I think there may be some minor misunderstandings.   Later
tomorrow my time i'll get to it -- it's bedtime here :-)

File: opencog/nlp/relex2logic/rule-helpers.scm (hunk 1 above)
Comment (rodsol) [inline opencog/nlp/relex2logic/rule-helpers.scm]: yes there is on my pull request https://github.com/opencog/relex/pull/223

**Comment by rodsol** [comment]
@williampma 


At the link you referred it says
“The SatisfyingSetLink is used to convert a predicate to a concept. That is, given a predicate that expresses truth, the SatisfyingSetLink is used to declare the set of all things that satisfy that predicate: that set becomes a new concept. “

I saw 2 usages of the SatisfyingSetLink in the current R2L rules 

1) Usage one: with variable node
Example:"Which girl do you like?" 

   (SatisfyingSetLink
         (VariableNode "$lKsNpXYtuyoVL7PjKJ6VqtTfVwKwJucmItZu")
         (EvaluationLink
            (PredicateNode "like@2bf6bc35-0350-4618-8c48-3ff9b225a0dd")
            (ListLink
               (ConceptNode "you@13866c5c-4255-4949-b638-fb86ac62a981")
               (VariableNode "$lKsNpXYtuyoVL7PjKJ6VqtTfVwKwJucmItZu")
            )
         )
      )
This agrees with the above definition: there is a predicate and anything that satisfies this predicate will be within the set the satisfyingset declares 

2) Usage two:
"He ate his lunch slowly."
_advmod(eat, slowly)

==>

(InheritanceLink
         (SatisfyingSetLink
            (PredicateNode "ate@eaf5f720-95e2-4011-9131-cc40a5796366")
         )
         (ConceptNode "slowly@9359c46e-fbac-4c12-8cfd-35488cd7ddab")
      )

This kinda makes sense for me semantically  
Here, There is also a predicate (PredicateNode "ate@eaf5f720-95e2-4011-9131-cc40a5796366") and anything that satisfies this predicate will be the new concept and the InheritanceLink tells that new concept is (ConceptNode "slowly@9359c46e-fbac-4c12-8cfd-35488cd7ddab")

Syntactically: Maybe as you said we need to add the variable node


But for my example in my pull request here is the current R2L output

"we shall meet at the library."
_advmod(meet, at)

(InheritanceLink
         (SatisfyingSetLink
            (PredicateNode "meet@09b0cfda-f832-4c6d-922d-f829f3f2db01")
         )
         (ConceptNode "at@b89b3e32-5db4-436a-842b-0f5ae17cd001")
      )

_pobj(at, library)

 (EvaluationLink
         (PredicateNode "at@b89b3e32-5db4-436a-842b-0f5ae17cd001")
         (ListLink
            (ConceptNode "library@d1885501-2ed3-46dd-ba5b-c94e969ee3d0")
         )
      )

I don't think this is correct: as you can see at the first output it makes the preposition “at” a ConceptNode. Prepositions are predicates, is that right? and then  it makes the preposition “at” a predicate on the second output

What I was trying to do was
First evaluate the 2 predicates and anything that satisfies the predicates would be declared as a set by using the satisfyingset which will be the new concepts 
Finally those concepts would be linked using InheriatanceLink 

(InheritanceLink
(SatisfyingSetLink
(PredicateNode "meet@123"))
(SatisfyingSetLink
(EvaluationLink
(PredicateNode "at@417")
(ListLink
(ConceptNode "library@661")))))

I don’t know if it makes sense for you, anyway what is your suggestion for this particular example?

**Comment by bgoertzel** [comment]
This is a functional use of satisfyingset as suggested by Nil

However, the pattern matcher likely does not support this yet; so, I
suppose variables should be inserted

Ben

**Comment by bgoertzel** [comment]
But ... hmmm... can't we do something like this? ...

****


Example: "we shall meet at the library."


_advmod(meet, at)
_pobj(at, library)

==>


         EvaluationLink

               PredicateNode "meet@123"

               ConceptNode "we@456"



         EvaluationLink

                   PredicateNode "at@417"

                   ListLink

                           ConceptNode "meet@123"

                           ConceptNode "library@661"




On Fri, Jul 24, 2015 at 8:11 AM, Ben Goertzel <ben@goertzel.org> wrote:

> This is a functional use of satisfyingset as suggested by Nil
>
> However, the pattern matcher likely does not support this yet; so, I
> suppose variables should be inserted
>
> Ben
> On Jul 24, 2015 8:08 AM, "Rodas Solomon" <notifications@github.com> wrote:
>
>> @williampma <https://github.com/williampma>
>>
>> At the link you referred it says
>> “The SatisfyingSetLink is used to convert a predicate to a concept. That
>> is, given a predicate that expresses truth, the SatisfyingSetLink is used
>> to declare the set of all things that satisfy that predicate: that set
>> becomes a new concept. “
>>
>> I saw 2 usages of the SatisfyingSetLink in the current R2L rules
>>
>> 1) Usage one: with variable node
>> Example:"Which girl do you like?"
>>
>> (SatisfyingSetLink
>> (VariableNode "$lKsNpXYtuyoVL7PjKJ6VqtTfVwKwJucmItZu")
>> (EvaluationLink
>> (PredicateNode "like@2bf6bc35-0350-4618-8c48-3ff9b225a0dd")
>> (ListLink
>> (ConceptNode "you@13866c5c-4255-4949-b638-fb86ac62a981")
>> (VariableNode "$lKsNpXYtuyoVL7PjKJ6VqtTfVwKwJucmItZu")
>> )
>> )
>> )
>> This agrees with the above definition: there is a predicate and anything
>> that satisfies this predicate will be within the set the satisfyingset
>> declares
>>
>> 2) Usage two:
>> "He ate his lunch slowly."
>> _advmod(eat, slowly)
>>
>> ==>
>>
>> (InheritanceLink
>> (SatisfyingSetLink
>> (PredicateNode "ate@eaf5f720-95e2-4011-9131-cc40a5796366")
>> )
>> (ConceptNode "slowly@9359c46e-fbac-4c12-8cfd-35488cd7ddab")
>> )
>>
>> This kinda makes sense for me semantically
>>
>> Here, There is also a predicate (PredicateNode
>> "ate@eaf5f720-95e2-4011-9131-cc40a5796366") and anything that satisfies
>> this predicate will be the new concept and the InheritanceLink tells that
>> new concept is (ConceptNode "slowly@9359c46e-fbac-4c12-8cfd-35488cd7ddab
>> ")
>>
>> Syntactically: Maybe as you said we need to add the variable node
>>
>> But for my example in my pull request here is the current R2L output
>>
>> "we shall meet at the library."
>> _advmod(meet, at)
>>
>> (InheritanceLink
>> (SatisfyingSetLink
>> (PredicateNode "meet@09b0cfda-f832-4c6d-922d-f829f3f2db01")
>> )
>> (ConceptNode "at@b89b3e32-5db4-436a-842b-0f5ae17cd001")
>> )
>>
>> _pobj(at, library)
>>
>> (EvaluationLink
>> (PredicateNode "at@b89b3e32-5db4-436a-842b-0f5ae17cd001")
>> (ListLink
>> (ConceptNode "library@d1885501-2ed3-46dd-ba5b-c94e969ee3d0")
>> )
>> )
>>
>> I don't think this is correct: as you can see at the first output it
>> makes the preposition “at” a ConceptNode. Prepositions are predicates, is
>> that right? and then it makes the preposition “at” a predicate on the
>> second output
>>
>> What I was trying to do was
>> First evaluate the 2 predicates and anything that satisfies the
>> predicates would be declared as a set by using the satisfyingset which will
>> be the new concepts
>> Finally those concepts would be linked using InheriatanceLink
>>
>> (InheritanceLink
>> (SatisfyingSetLink
>> (PredicateNode "meet@123"))
>> (SatisfyingSetLink
>> (EvaluationLink
>> (PredicateNode "at@417")
>> (ListLink
>> (ConceptNode "library@661")))))
>>
>> I don’t know if it makes sense for you, anyway what is your suggestion
>> for this particular example?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/opencog/opencog/pull/1691#issuecomment-124339377>.
>>
>


-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by williampma** [comment]
That looks nice and compact, but what does it mean for "meet@123" to be both a ConceptNode and a PredicateNode?

**Comment by bgoertzel** [comment]
Oops, thanks ;) ...

***

         EvaluationLink

               PredicateNode "meet@123"

               ConceptNode "we@456"



         EvaluationLink

                   PredicateNode "at@417"

                   ListLink

                           PredicateNode "meet@123"

                           ConceptNode "library@661"

On Fri, Jul 24, 2015 at 10:45 AM, William Ma <notifications@github.com>
wrote:

> That looks nice and compact, but what does it mean for "meet@123" to be
> both a ConceptNode and a PredicateNode?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/pull/1691#issuecomment-124445090>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by ruiting** [comment]
I didn't trace the current criteria of the R2L rules recently... The
previous way of representing this kind of sentence is like:

Example: "we shall meet at the library."
_advmod(meet, at)
_pobj(at, library)

==>
(EvaluationLink
     (PredicateNode "meet@123")
     (ListLink
           (ConceptNode "We@256")
           (EvaluationLink
                 (PredicateNode "at@417")
                 (ListLink
                      (ConceptNode "library@661")))))

Ruiting Lian

**Comment by ruiting** [comment]
oops, didn't see Ben's reply before sending it

Ruiting Lian

**Comment by anitzkin** [comment]
Okay . . . excuse me if idon't address all of what's been saidin this
thread; it's easier to cut to the simplest points . . .

-- I don't know why you would want to combine the prepositional-phrase-rule
and the advmod-rule into one adverbialpp rule.  I seperated them for a good
reason.  It allows much more flexibility and gernerativity in the rule
base.  A prepostiional phrase can serve a variety of roles and a variety of
things can play the role of an adverbial.  Why not leave them as two
seperate rules?

-- Seems like the best solution to the preposition as concept versus
predicate node is pretty much what Ben suggested except done by two rules:

_advmod(meet, at)

EvaluationLink
     PredicateNode "at@123"
          ListLink
                PredicateNode "meet"
and

_pobj(at, library)

EvaluationLink
     Predicatenode "at@123"
              ListLink
                      ConceptNode "library"

This would imply retaining the general advmod rule, in which prepositions
and compleentizers may also play the role of the adverb, but changing the
rule so that all adverbs and adverbials get treated as in the example above
instead of using inheritance.  This makes sense to me.  Of course, you
could still be free to have a separate rule for adverbs versus adverbials,
and treat adverbs with inheritance and adverbial phrases as above.

**Comment by rodsol** [comment]
@bgoertzel 
"He ate his lunch slowly."
_advmod(eat, slowly)

==>

(InheritanceLink
            (SatisfyingSetLink
               (PredicateNode "ate@123"))
            (ConceptNode "slowly@345"))

do you think this is correct ?

**Comment by ruiting** [comment]
I remember the discussion about it. This is correct. As inheritanceLink
can't connect PredicateNode with ConceptNode, but it's logically correct,
so we need to use SatisyingSet to make it grammatical

**Comment by bgoertzel** [comment]
Yeah, it's correct, using the functional style for SatisfyingSet as
advocated by Nil ...

On Sat, Jul 25, 2015 at 8:03 AM, Rodas Solomon <notifications@github.com>
wrote:

> @bgoertzel <https://github.com/bgoertzel>
> "He ate his lunch slowly."
> _advmod(eat, slowly)
>
> ==>
>
> (InheritanceLink
> (SatisfyingSetLink
> (PredicateNode "ate@123"))
> (ConceptNode "slowly@345"))
>
> do you think this is correct ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/pull/1691#issuecomment-124806282>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by rodsol** [comment]
@ruiting 
OK, but still we need to add the variablenode for satisfyingset , right?

**Comment by bgoertzel** [comment]
Rodi, that depends on the PLN rules.  I think would be desirable to make
the PLN rules handle the functional form of SatisfyingSet also.   This
doesn't seem very hard...

I.e.

SatisfyingSet
    PredicateNode $P

is equivalent to

SatisfyingSet
    VariableNode $X
    EvaluationLink
          PredicateNode $P
          VariableNode $X

and we could just create an inference rule enacting this equivalence...
which would enable encodings to be done using the functional format..

ben

On Sat, Jul 25, 2015 at 10:16 AM, Rodas Solomon <notifications@github.com>
wrote:

> @ruiting <https://github.com/ruiting>
> OK, but still we need to add the variablenode for satisfyingset , right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/pull/1691#issuecomment-124819843>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by rodsol** [comment]
alright

**Pull request merged**